### ISSUE TITLE
AES GCM 

### ISSUE THREAD

**Issue by alfaro28** [Legrandin/pycryptodome#195]
I getting different results depending on the machine I'm running my code, this is my function

    def message_encrypt_sign(message, key, nonce=None):
        message = message.encode("utf-8")
        key = bytes.fromhex(key)

        if nonce is None:
            b = get_random_bytes(16)
            nonce = b.hex()

        nonce = bytes.fromhex(nonce)

        cipher = AES.new(key, AES.MODE_GCM, nonce)
        ciphertext, mac = cipher.encrypt_and_digest(message)

        ciphertext = base64.b64encode(ciphertext).decode('utf-8')
        mac = base64.b64encode(mac).decode('utf-8')
        nonce = nonce.hex()

        return ciphertext, mac, nonce

doing

`
message_encrypt_sign("this is a test string just to prove my point", "cba5b34b9a9e960c6eb69c1a7f31a36e5ef5f2131c7a052e1c7a91d955067f67", "496c6353686f4c41325a4f4865524675")
`

is giving me 

`
('NL1MvwIGnUZbKNW/E/TZFPqUYlXqNRFWEGWMgCV1kOK0V2hCYrgfTbPF7qE=', 'yKTVdJUpKKZY6DOQjMrW0g==', '496c6353686f4c41325a4f4865524675')
`

on my local machine and 

`
('0mlf47gmOn6h+HZg/g/h2zKkmB9OPUwWKfUn0gzoU7IL7N3aQIajSW4KlhE=', 'Q4XuY0m/J6rj9O3OGCXU2Q==', '496c6353686f4c41325a4f4865524675')
`

on an AWS EC2 instance, any ideas why?

Edit:

On my machine no matter what it always gives me the same result, on the server it gives a different result everytime I launch a diferent interpreter

**Comment by Legrandin** [comment]
What happens on the server is clearly dodgy.
It looks like that the GCM cipher is actually using a random nonce, instead of the one you provide.
Can you add a line `assert cipher.nonce.hex() == "496c6353686f4c41325a4f4865524675"` right after `cipher.new`, and then call `message_encrypt_sign` with all of the three values?

**Comment by alfaro28** [comment]
I changed my function to 

```
def message_encrypt_sign(message, key, nonce=None):
    message = message.encode("utf-8")
    key = bytes.fromhex(key)

    if nonce is None:
        b = get_random_bytes(16)
        nonce = b.hex()

    nonce = bytes.fromhex(nonce)
    cipher = AES.new(key, AES.MODE_GCM, nonce)

    assert cipher.nonce.hex() == "496c6353686f4c41325a4f4865524675"

    ciphertext, mac = cipher.encrypt_and_digest(message)
    ciphertext = base64.b64encode(ciphertext).decode('utf-8')
    mac = base64.b64encode(mac).decode('utf-8')

    return ciphertext, mac, cipher.nonce.hex()
```

and still get the same behavior

**Comment by Legrandin** [comment]
What versions do you have on the EC2 instance (`pip show pycryptodome`, Python version, OS)?

**Comment by alfaro28** [comment]
pycryptodomex==3.6.4

Python 3.6.2

OS: Amazon Linux AMI release 2017.09

**Comment by Legrandin** [comment]
I am running this on EC2:

```python
import sys
import base64
import Cryptodome
from Cryptodome.Cipher import AES

def message_encrypt_sign(message, key, nonce=None):
    message = message.encode("utf-8")
    key = bytes.fromhex(key)

    if nonce is None:
        b = get_random_bytes(16)
        nonce = b.hex()

    nonce = bytes.fromhex(nonce)
    cipher = AES.new(key, AES.MODE_GCM, nonce)

    assert cipher.nonce.hex() == "496c6353686f4c41325a4f4865524675"

    ciphertext, mac = cipher.encrypt_and_digest(message)
    ciphertext = base64.b64encode(ciphertext).decode('utf-8')
    mac = base64.b64encode(mac).decode('utf-8')

    return ciphertext, mac, cipher.nonce.hex()

print("Python version:", sys.version_info[:3])
print("Pycryptodomex version:", Cryptodome.__version__)
res = message_encrypt_sign("this is a test string just to prove my point", "cba5b34b9a9e960c6eb69c1a7f31a36e5ef5f2131c7a052e1c7a91d955067f67", "496c6353686f4c41325a4f4865524675")
print(res)
```

but I do get the intended output, every time:

```
Python version: (3, 6, 5)
Pycryptodomex version: 3.6.4
('NL1MvwIGnUZbKNW/E/TZFPqUYlXqNRFWEGWMgCV1kOK0V2hCYrgfTbPF7qE=', 'yKTVdJUpKKZY6DOQjMrW0g==', '496c6353686f4c41325a4f4865524675')
```

It is a newer AWS AMI, though 2018.03.

**Comment by Legrandin** [comment]
Could you please try with the new version of the library (3.6.5)? There is a fix for GCM mode which might be relevant for this case.

**Comment by alfaro28** [comment]
It seems to be fixed with the new version (3.6.5), its giving me the correct values now

**Comment by Legrandin** [comment]
Awesome! Thanks for checking it again!
